### ISSUE TITLE
Incorrect result after an INSERT violates a UNIQUE constraint

### ISSUE THREAD

**Issue by mrigger** [cwida/duckdb#514]
Consider the following statements:

```sql
CREATE TABLE t0(c0 INT);
CREATE UNIQUE INDEX i0 ON t0(c0);
INSERT INTO t0(c0) VALUES (1);
INSERT INTO t0(c0) VALUES (1); -- Failed to commit: Constraint: PRIMARY KEY or UNIQUE constraint violated: duplicated key
SELECT * FROM t0 WHERE t0.c0 = 1; -- expected: {1}, actual: {}
```
Unexpectedly, the `SELECT` does not fetch a row. There is one row c0=1 present in the table. It seems that the second, failing `INSERT` triggers this bug; if this statement is removed, the query works as expected.

Detailed output:
```duckdb/build/release$ ./duckdb_cli 
SQLite version DuckDB c52fc9ba
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
sqlite> CREATE TABLE t0(c0 INT);
sqlite> CREATE UNIQUE INDEX i0 ON t0(c0);
sqlite> INSERT INTO t0(c0) VALUES (1);
sqlite> INSERT INTO t0(c0) VALUES (1);
Error: TransactionContext: Failed to commit: Constraint: PRIMARY KEY or UNIQUE constraint violated: duplicated key
sqlite> SELECT * FROM t0 WHERE t0.c0 = 1; -- expected: {1}, actual: {}
sqlite> 
```

I can reproduce this on the latest master (commit c52fc9bafd22e41a5554d3305530ebc1bf8364f9) and the current fix branch (commit 5c4cde52e3c3ddc4ad4d78a7adf0101f7e61861f).

**Comment by Mytherin** [comment]
This is likely related to an incorrect cleanup of the index structure after the transaction abort; looks to me like the row is removed from the index where it shouldn't be.

**Comment by Mytherin** [comment]
Fixed in 0d29535fe6aefdf2fbe8a7de0775c253f8038a1a, the problem was that when the transaction failed to commit it would try to remove entries from the index again. This behavior is there so the index is properly maintained if a conflict occurs later on, e.g. if we have `INSERT (1), (2), (3)`, with a conflict on (3), we want to remove (1) and (2) again from the index.

However, before it performed this cleanup only by value, and did not correctly consider the row id. This caused the first row to be incorrectly removed from the index in the cleanup of the second insertion.
